G35 assisted tramming improvements

qwewer0 opened the pull request:
### Description

- Adding `ASSISTED_TRAMMING_WAIT_POSITION` to `G35`.
    This already existed, but worked only with the tramming wizard and not the G35 command.

- Removing `ASSISTED_TRAMMING_MENU_ITEM`. 
    (https://github.com/MarlinFirmware/Marlin/pull/20298#discussion_r549153020)
    This only ran G35, which provides no value without being able to see the serial output.
    Users wishing for an interactive experience should use the wizard option.

- Disables homing Z after the G35 comment
    There was no value in homing Z after executing G35, because at that point you have not yet adjusted the screws.
    Instead, simply invalidate Z, so that homing is required before next use.

### Benefits

Easier screw adjustment after `G35` is done, in particular for printers where the print head blocks screw access.
<!-- What does this fix or improve? -->

### Related Issues

https://github.com/MarlinFirmware/Marlin/issues/20281
<!-- Whether this fixes a bug or fulfills a feature request, please list any related Issues here. -->

qwewer0 committed:
Update G35.cpp

Adding ASSISTED_TRAMMING_WAIT_POSITION to G35

qwewer0 committed:
Update menu_tramming.cpp

ASSISTED_TRAMMING_WAIT_POSITION Comment Consistency

qwewer0 committed:
Update G35.cpp

Z axis is alreagy set to be unhomed.

qwewer0 committed:
Simplify repetition

qwewer0 committed:
move_to_wait_pos

qwewer0 committed:
Comment_change

qwewer0 committed:
wait_pos_move-move_to_tramming_wait_pos

thinkyhead committed:
Add test, tramming.cpp

swissnorp commented on Marlin/src/gcode/bedlevel/G35.cpp:
```diff
@@ -178,11 +158,11 @@ void GcodeSuite::G35() {
-
-  // Home Z after the alignment procedure
-  process_subcommands_now_P(PSTR("G28Z"));
 }
```
I think we should keep homing after the wizard. @qwewer0 seems you dont like homing 😄. We may could do it like as its handled in the probe offset wizard:
```
  // If probe is used for homing and there is no fixed endstop, rehome Z
  #if EITHER(Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN, USE_PROBE_FOR_Z_HOMING)
    set_axis_never_homed(Z_AXIS);
    queue.inject_P(PSTR("G28Z"));
  #endif

  // Move XYZ to wait position
  move_to_tramming_wait_pos();
```

I'd like to see a printer homing after a calibration/bedleveling process if it changes either the height of the bed relative to the nozzle or the nozzle to probe offset. It sets a printer in a ready state and is done relaitvely fast and there for more convenient then going back to the menu for homing all axes instead of Z only (if individual homing is deactivated).

I would consider to make a change some lines above (beginning of the percedure) from `home_all_axes();` to `queue.inject_P(PSTR("G28O"));`

qwewer0 replied:
My idea to why not to home after the G35 is, because after the bed adjustment we would home the printer either way when printing a gcode file, or doing another G35 which home too at the start, so homing twice would be a waste of time.

qwewer0 replied:
But I'm still open in the idea, just not convinced...

qwewer0 replied:
> I would consider to make a change some lines above (beginning of the percedure) from home_all_axes(); to queue.inject_P(PSTR("G28O"));

With small changes on the screws the bed center could stay within the `Z_PROBE_LOW_POINT` area, but if the bed was badly tilted, it isn't guaranteed that the middle of the bed would stay within the `Z_PROBE_LOW_POINT` area. And we cannot be sure that homing position is at the center of the bed.
So I don't think it would be a good idea to change the first home to an optional one, because if the previous bed adjustment after the G35 is too great, then the `Z_PROBE_LOW_POINT` would trigger and give use an error.
Even if I'm wrong in this, I still think the first home should stay there...

thinkyhead replied:
While inside of a G-code command no other G-code commands in the queue will execute. So you would never want to use `queue.inject_P` to do something that you want to happen before your currently-processing G-code exits. Instead, you should use `process_subcommands_now_P` for commands that you want be executed here and now.

swissnorp replied:
> 
> 
> > I would consider to make a change some lines above (beginning of the percedure) from home_all_axes(); to queue.inject_P(PSTR("G28O"));
> 
> With small changes on the screws the bed center could stay within the `Z_PROBE_LOW_POINT` area, but if the bed was badly tilted, it isn't guaranteed that the middle of the bed would stay within the `Z_PROBE_LOW_POINT` area. And we cannot be sure that homing position is at the center of the bed.
> So I don't think it would be a good idea to change the first home to an optional one, because if the previous bed adjustment after the G35 is too great, then the `Z_PROBE_LOW_POINT` would trigger and give use an error.
> Even if I'm wrong in this, I still think the first home should stay there...

I agree 👍 

And I stay with my opinion and suggest the following changes for the end of the tramming process:

```
  // If probe is used for homing and there is no fixed endstop, rehome Z
  #if HOMING_Z_WITH_PROBE
    set_axis_never_homed(Z_AXIS);
    process_subcommands_now_P(PSTR("G28Z"));
  #endif

  // Move XYZ to wait position
  move_to_tramming_wait_pos();
```

swissnorp commented on Marlin/src/gcode/bedlevel/G35.cpp:
```diff
@@ -178,11 +158,11 @@ void GcodeSuite::G35() {
   // the probe deployed if it was successful.
   probe.stow();
 
+  // Move XYZ to wait position
+  move_to_tramming_wait_pos();
+
   // After this operation the Z position needs correction
   set_axis_never_homed(Z_AXIS);
-
-  // Home Z after the alignment procedure
-  process_subcommands_now_P(PSTR("G28Z"));
```
@qwewer0 
> 
> 
> > I would consider to make a change some lines above (beginning of the percedure) from home_all_axes(); to queue.inject_P(PSTR("G28O"));
> 
> With small changes on the screws the bed center could stay within the `Z_PROBE_LOW_POINT` area, but if the bed was badly tilted, it isn't guaranteed that the middle of the bed would stay within the `Z_PROBE_LOW_POINT` area. And we cannot be sure that homing position is at the center of the bed.
> So I don't think it would be a good idea to change the first home to an optional one, because if the previous bed adjustment after the G35 is too great, then the `Z_PROBE_LOW_POINT` would trigger and give use an error.
> Even if I'm wrong in this, I still think the first home should stay there...

I agree 👍 

And I stay with my opinion for the end of the precedure. I would like to see a rehoming as follow:

```
  // If probe is used for homing and there is no fixed endstop, rehome Z
  #if HOMING_Z_WITH_PROBE
    set_axis_never_homed(Z_AXIS);
    process_subcommands_now_P(PSTR("G28Z"));
  #endif

  // Move XYZ to wait position
  move_to_tramming_wait_pos();
```

qwewer0 replied:
I still don't think that there is any reason to home after G35, because at the end the z axis will be set to be unhomed (`set_axis_never_homed(Z_AXIS)`).

swissnorp replied:
Well at least I think `set_axis_never_homed(Z_AXIS)` should only be done if `HOMING_Z_WITH_PROBE` is true. Or are there printers out there, which have the z endstop moving with bed adjustment up and down?

I think it would be more convenient to have it homed. Especially because it's Z only. But to have consistency over all tramming / level bed corners options I think it okey to just do:

```
  // If probe is used for homing meassuring the bed heigth, set Z unhomed
  #if HOMING_Z_WITH_PROBE
    set_axis_never_homed(Z_AXIS);
  #endif
```
 And i think we should add this also to `Marlin/src/lcd/menu/menu_tramming.cpp`

```
void tramming_wizard_menu() {
  DEBUG_ECHOLNPAIR("Screen: tramming_wizard_menu");
  START_MENU();
  STATIC_ITEM(MSG_SELECT_ORIGIN);
  // Draw a menu item for each tramming point
  LOOP_L_N(i, G35_PROBE_COUNT)
    SUBMENU_N_P(i, (char*)pgm_read_ptr(&tramming_point_name[i]), []{ _menu_single_probe(MenuItemBase::itemIndex); });
  ACTION_ITEM(MSG_BUTTON_DONE, []{ 
    #if HOMING_Z_WITH_PROBE  // If probe is used for homing by referencing the bed heigth, set Z unhomed
      set_axis_never_homed(Z_AXIS);
    #endif
    ui.goto_previous_screen_no_defer();
  });
  END_MENU();
}
```

qwewer0 replied:
> Well at least I think `set_axis_never_homed(Z_AXIS)` should only be done if `HOMING_Z_WITH_PROBE` is true. Or are there printers out there, which have the z endstop moving with bed adjustment up and down?

Not sure why should `set_axis_never_homed(Z_AXIS)` depend on `HOMING_Z_WITH_PROBE`. It only sets the axis untrasted and unhomed, so that the next homing must happen, but the stepper will still be engaged, at least I believe so.
I'm a bit lost on your second sentence, could you explain it a bit more?

> I think it would be more convenient to have it homed. Especially because it's Z only. But to have consistency over all tramming / level bed corners options I think it okey to just do:
> 
> ```
>   // If probe is used for homing meassuring the bed heigth, set Z unhomed
>   #if HOMING_Z_WITH_PROBE
>     set_axis_never_homed(Z_AXIS);
>   #endif
> ```
> 
> And i think we should add this also to `Marlin/src/lcd/menu/menu_tramming.cpp`

One of my problem with the homing at the end is that it isn't just a Z homing, but a full homing sequence, at least for me with a BLTouch, and it takes "way" too much time, and for what? After the bed adjustment we cannot trust the previously measured homing height, and that is why there is the `set_axis_never_homed(Z_AXIS)` at the end.

swissnorp replied:
> Not sure why should `set_axis_never_homed(Z_AXIS)` depend on `HOMING_Z_WITH_PROBE`. It only sets the axis untrasted and unhomed, so that the next homing must happen, but the stepper will still be engaged, at least I believe so.
> I'm a bit lost on your second sentence, could you explain it a bit more?

If we have homing with an endstop mounted to the gantry, there is no need to set Z untrusted because it's still at the same `current_position.z` relative to `Z_HOME_POS`. So it's only needed to go for `set_axis_never_homed(Z_AXIS);` if `HOMING_Z_WITH_PROBE`.

> One of my problem with the homing at the end is that it isn't just a Z homing, but a full homing sequence, at least for me with a BLTouch, and it takes "way" too much time, and for what? After the bed adjustment we cannot trust the previously measured homing height, and that is why there is the `set_axis_never_homed(Z_AXIS)` at the end.

`G28Z ` will home only Z even with BLTouch, because X and Y are known/trusted. I don't know your config, but am I right that you use an Ender3? (Same as me, also with BLTouch). My Printer needs max about 6 to 7 seconds to do G28 Z.

qwewer0 replied:
> If we have homing with an endstop mounted to the gantry, there is no need to set Z untrusted because it's still at the same `current_position.z` relative to `Z_HOME_POS`. So it's only needed to go for `set_axis_never_homed(Z_AXIS);` if `HOMING_Z_WITH_PROBE`.

By my understanding `set_axis_never_homed(Z_AXIS)` is needed because the Z position changes relative to the endstop/probe, but the bed height changes, and by that the Z position relative to the bed. If the Z axis stays trusted at the end and the bed adjustment was relatively large, then on the next Z movement to zero, the nozzle would grind into the bed, which is why `set_axis_never_homed(Z_AXIS)` is needed for both endstop and probe.

> `G28Z ` will home only Z even with BLTouch, because X and Y are known/trusted. I don't know your config, but am I right that you use an Ender3? (Same as me, also with BLTouch). My Printer needs max about 6 to 7 seconds to do G28 Z.

You are right, `process_subcommands_now_P(PSTR("G28Z"))` does only move to `Z_SAFE_HOMING` position then only homes Z, but even if it would take less than a second I still think that it is just an unnecessary extra movement, as `set_axis_never_homed(Z_AXIS)` needs to be at the end, because of what is written above.

swissnorp replied:
> By my understanding `set_axis_never_homed(Z_AXIS)` isn't needed because the Z position changes relative to the endstop/probe, but the bed height changes, and by that the Z position relative to the bed.

**isn't?** typo?

Z position **only** changes relative to the bed, because the bed is moved/adjusted!

>If the Z axis stays trusted at the end and the bed adjustment was relatively large, then on the next Z movement to zero, the nozzle would grind into the bed, which is why `set_axis_never_homed(Z_AXIS)` is needed for both endstop and probe.

This is only valid for a probe/endstop that touches the bed for homing! And therefor `HOMING_Z_WITH_PROBE`.

sjasonsmith replied:
Unresolving this comment, since it seems important to see this discussion during review.

qwewer0 replied:
> **isn't?** typo?

Sorry, it should be "is", edited the comment. Thank you for catching it.

> Z position **only** changes relative to the bed, because the bed is moved/adjusted!

Yes, that is the problem. ⬇️ 

> This is only valid for a probe/endstop that touches the bed for homing! And therefor `HOMING_Z_WITH_PROBE`.

If the bed would only be adjusted downwards, then it wouldn't be a problem, but that isn't always the case. If the bed is adjusted upwards, then the previously measured z position will be wrong and unusable and potentially dangerous if used.

qwewer0 replied:
> Unresolving this comment, since it seems important to see this discussion during review.

You are right, thank you.

sjasonsmith replied:
I agree that invalidating Z would only help if `HOMING_Z_WITH_PROBE`. If you are homing with the endstop it will return to the same position, because endstops don't measure the bed surface!

sjasonsmith replied:
I also agree that there is no need to actually home _anything_ after the G35 command. There are two possible outcomes from G35, and homing makes sense in neither of them.

1. No screw adjustment needed.

2. Screw adjustment needed, **but they aren't adjusted yet!**. There is no point to homing Z, before the screws are actually adjusted!

qwewer0 replied:
<s>@sjasonsmith Do you think this needs other opinions or can be set as resolved?<s/>

qwewer0 replied:
> I agree that invalidating Z would only help if `HOMING_Z_WITH_PROBE`. If you are homing with the endstop it will return to the same position, because endstops don't measure the bed surface!

I think that it cannot hurt to have z invalidation on a z-endstop setup too and to know over time that it is useless or not, but it would be an easy fix:
```
  // If probe is used for homing, set Z axis never homed
  #if HOMING_Z_WITH_PROBE
    set_axis_never_homed(Z_AXIS);
  #endif
```

sjasonsmith replied:
I don’t feel too strongly about it. You are already removing the actual homing at the end of the routine. We can wait and see if people still complain.

sjasonsmith commented:
@qwewer0 & @swissnorp it seems you are still working through some issues? Please reply when you have things worked out and think this is ready to review.

qwewer0 commented:
> @qwewer0 & @swissnorp it seems you are still working through some issues? Please reply when you have things worked out and think this is ready to review.

@sjasonsmith I think this is ready for review.
I tested it and it is working as intended.

sjasonsmith commented:
Sorry, this got a little lost in the pile and I never came back to it. Unfortunately I've now posted other changes which will conflict with this!

At this point I would prefer to wait for #20572 to be merged, then I can come back and resolve any issues with this PR.

sjasonsmith commented:
I actually ran across this missing capability when I was testing that other PR. It was completely non-obvious in the config file that it applied only to the wizard!

sjasonsmith commented on Marlin/src/inc/SanityCheck.h:
```diff
@@ -2156,8 +2156,12 @@ static_assert(hbm[Z_AXIS] >= 0, "HOMING_BUMP_MM.Z must be greater than or equal
+#if ENABLED(ASSISTED_TRAMMING)
+  #if !HAS_BED_PROBE
+    #error "ASSISTED_TRAMMING requires a bed probe."
+  #elif BOTH(ASSISTED_TRAMMING_MENU_ITEM, ASSISTED_TRAMMING_WIZARD)
+    #error "Please select either ASSISTED_TRAMMING_MENU_ITEM or ASSISTED_TRAMMING_WIZARD, not both."
+  #endif
```
This is better than the undocumented current behavior.

Maybe we should instead just allow both to be enabled at the same time? They are somewhat redundant, but still different...

qwewer0 replied:
We could have both, but I don't really understand why do we even have `ASSISTED_TRAMMING_MENU_ITEM` in general as that is only a `G35` gcode button, and without Pronterface, Octoprint etc. there is no use to it.

qwewer0 replied:
If we have to use Pronterface, Octoprint etc. anyway for `G35`, then `ASSISTED_TRAMMING_MENU_ITEM` has no place in the printer menu in my opinion.
Approaching the problem from a different perspective, why do we have a menu item that is unusable with only the printer.

sjasonsmith replied:
I think you are right. If it doesn’t report information back to the screen it doesn’t provide any real value. It seems unlikely that people would want to monitor it over serial, but don’t also have a way to send G35 over serial.

qwewer0 committed:
Remove ASSISTED_TRAMMING_MENU_ITEM

"If it doesn’t report information back to the screen it doesn’t provide any real value. It seems unlikely that people would want to monitor it over serial, but don’t also have a way to send G35 over serial."

qwewer0 committed:
SanityCheck condition change

qwewer0 commented:
On the latest Pronterface they fixed the issue (https://github.com/kliment/Printrun/issues/1103) with Clockwise, Counter-Clockwise not showing up correctly. Should we change it back or leave it as it is right now?
(https://github.com/MarlinFirmware/Marlin/commit/d02d7a95d13e27e0f90d2226c14d103b4e5cde3c#diff-3a120a4cc3437ddf190a5eb0a767a5c13a907296d9929e9edce30886cef9e3e9)

sjasonsmith committed:
Minor cleanup, make functions static

sjasonsmith commented:
> Should we change it back or leave it as it is right now?

I don't know, but I don't think this change should be dependent on that.

qwewer0 commented on Marlin/src/lcd/menu/menu_tramming.cpp:
```diff
@@ -99,4 +96,4 @@ void goto_tramming_wizard() {
 }
 
-#endif // HAS_LCD_MENU && ASSISTED_TRAMMING_WIZARD
+#endif // BOTH(HAS_LCD_MENU, ASSISTED_TRAMMING_WIZARD)
```
I don't think a comment needs that change.

sjasonsmith replied:
It really doesn't matter either way. I don't like closing comments at all, but if we are going to have them I think they should match.

qwewer0 replied:
Are you planning to change all the comments in marlin to that too? Because I only see one like that:
https://github.com/MarlinFirmware/Marlin/blob/4402a0578a9fa5c9743eb5774224ce206c618ce9/Marlin/src/lcd/extui/lib/ftdi_eve_touch_ui/screens/leveling_menu.cpp#L121

qwewer0 commented on Marlin/src/lcd/menu/menu_tramming.cpp:
```diff
@@ -42,20 +42,18 @@
   DEBUG_ECHOLNPAIR("probe_single_point: ", z_probed_height, "mm");
   z_measured[tram_index] = z_probed_height;
-
-  // Move XYZ to wait position
```
Shouldn't we keep the comments?

sjasonsmith replied:
The function name is descriptive enough that the comment is redundant.

sjasonsmith reviewed (approved):
I took the liberty of updating the title/description to make the reasoning for the changes apparent at a glance.

I think these are great changes. These items were all confusing while I was working on the code to validate the selected tramming points at build-time.

Aside from making G35 more useful, I think they will also help reduce some new-user confusion.

qwewer0 committed:
Remove remaining wait comment

sjasonsmith commented:
@qwewer0 could you add a sanity check for the removed option?

qwewer0 committed:
Add sanitycheck for removed ASSISTED_TRAMMING_MENU_ITEM

qwewer0 commented:
> @qwewer0 could you add a sanity check for the removed option?

Done, could you check it, please?

thinkyhead committed:
Update menu_tramming.cpp

thinkyhead committed:
cleanup

qwewer0 commented:
@thinkyhead Thanks for the cleanup, looks good and logical. 👍

o5 commented:
Cannot build after rebase with this change :/ 

```bash
Compiling .pio/build/STM32F103RC_btt_USB/src/src/gcode/bedlevel/abl/M421.cpp.o
In file included from Marlin/src/feature/tramming.cpp:27:0:
Marlin/src/feature/tramming.h:36:36: error: static assertion failed: TRAMMING_POINT_XY point 0 is not reachable with the default NOZZLE_TO_PROBE offset and PROBING_MARGIN.
 #define VALIDATE_TRAMMING_POINT(N) static_assert(N >= G35_PROBE_COUNT || Probe::build_time::can_reach(screws_tilt_adjust_pos[N]), \
                                    ^
Marlin/src/feature/tramming.h:38:1: note: in expansion of macro 'VALIDATE_TRAMMING_POINT'
 VALIDATE_TRAMMING_POINT(0); VALIDATE_TRAMMING_POINT(1); VALIDATE_TRAMMING_POINT(2); VALIDATE_TRAMMING_POINT(3); VALIDATE_TRAMMING_POINT(4);
 ^~~~~~~~~~~~~~~~~~~~~~~
Marlin/src/feature/tramming.h:36:36: error: static assertion failed: TRAMMING_POINT_XY point 3 is not reachable with the default NOZZLE_TO_PROBE offset and PROBING_MARGIN.
 #define VALIDATE_TRAMMING_POINT(N) static_assert(N >= G35_PROBE_COUNT || Probe::build_time::can_reach(screws_tilt_adjust_pos[N]), \
                                    ^
Marlin/src/feature/tramming.h:38:85: note: in expansion of macro 'VALIDATE_TRAMMING_POINT'
 VALIDATE_TRAMMING_POINT(0); VALIDATE_TRAMMING_POINT(1); VALIDATE_TRAMMING_POINT(2); VALIDATE_TRAMMING_POINT(3); VALIDATE_TRAMMING_POINT(4);
                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~
*** [.pio/build/STM32F103RC_btt_USB/src/src/feature/tramming.cpp.o] Error 1
In file included from Marlin/src/gcode/bedlevel/G35.cpp:43:0:
Marlin/src/gcode/bedlevel/../../feature/tramming.h:36:36: error: static assertion failed: TRAMMING_POINT_XY point 0 is not reachable with the default NOZZLE_TO_PROBE offset and PROBING_MARGIN.
 #define VALIDATE_TRAMMING_POINT(N) static_assert(N >= G35_PROBE_COUNT || Probe::build_time::can_reach(screws_tilt_adjust_pos[N]), \
                                    ^
Marlin/src/gcode/bedlevel/../../feature/tramming.h:38:1: note: in expansion of macro 'VALIDATE_TRAMMING_POINT'
 VALIDATE_TRAMMING_POINT(0); VALIDATE_TRAMMING_POINT(1); VALIDATE_TRAMMING_POINT(2); VALIDATE_TRAMMING_POINT(3); VALIDATE_TRAMMING_POINT(4);
 ^~~~~~~~~~~~~~~~~~~~~~~
Marlin/src/gcode/bedlevel/../../feature/tramming.h:36:36: error: static assertion failed: TRAMMING_POINT_XY point 3 is not reachable with the default NOZZLE_TO_PROBE offset and PROBING_MARGIN.
 #define VALIDATE_TRAMMING_POINT(N) static_assert(N >= G35_PROBE_COUNT || Probe::build_time::can_reach(screws_tilt_adjust_pos[N]), \
                                    ^
Marlin/src/gcode/bedlevel/../../feature/tramming.h:38:85: note: in expansion of macro 'VALIDATE_TRAMMING_POINT'
 VALIDATE_TRAMMING_POINT(0); VALIDATE_TRAMMING_POINT(1); VALIDATE_TRAMMING_POINT(2); VALIDATE_TRAMMING_POINT(3); VALIDATE_TRAMMING_POINT(4);
```

Unfortunately, I don't understand completely this feature so I'm not able to fix assertion.  Could someone help me? There is my sources [marlin](https://github.com/diy4fun/marlin/tree/bugfix-2.0.x-latest), [marlin-configurations](https://github.com/diy4fun/marlin-configurations/tree/bugfix-2.0.x-latest).

qwewer0 commented:
@o5 https://github.com/MarlinFirmware/Marlin/pull/20572

o5 commented:
OK, I discovered that I enabled this option when reading some Reddit guide and (as most of others options... in that time, really!) and I didn't exactly know what I did  :) So time to change it! 

Looked into documentation and some explanation in youtube video but still cannot confidently answer my question: **Is this calibration necessary for me if I'm using auto bed leveling with fixed probe before every print using G29?** I think I should comment out `ASSISTED_TRAMMING`.

qwewer0 commented:
@o5 G35 can help you tram your bed. It is up to you if you want to use it or not, you can tram your bed with the paper method too.
